### DISCUSSION TITLE
Action::downloadURL does not download file (loads in-browser)

### DISCUSSION THREAD

**Discussion by miagg** [laravel/nova-issues#6299] [Ideas & Feature Requests] [answered]
- Laravel Version: 10.48.3
- Nova Version: 4.33.0
- PHP Version: 8.3.3
- Database Driver & Version: MySQL 8
- Operating System and Version: macOS Sonoma 14.3.1 (Laravel Herd)
- Browser type and version: Chrome 122.0

### Description:
`downloadUrl` static action shows image inside the browser instead of downloading it. I've also tried with Safari & Firefox.

### Detailed steps to reproduce the issue on a fresh Nova installation:
Add static action on any resource
```php
public function actions(Request $request): array
{
    return [
        Action::downloadURL('Download Image', 'http://laravel.test/test.png')->withoutConfirmation();
    ];
}
```

I used `withoutConfirmation()` to skip the dialog. I think it should be skipped by default for downloads anyway.
Expected behavior is to download the file. However I get the image displayed inside the browser.

**Comment by crynobone** [comment]
This functionality still depends on browser's default behavior.

**Reply by webard** [reply]
@crynobone browser's behavior depends on "download" attribute on `<a>` tag.

```
<a href="/images/myw3schoolsimage.jpg" download="downloaded_filename">
```

I think this attribute should be configurable via ```Action::downloadURL``` , it's value too.

**Reply by crynobone** [reply]
![CleanShot 2024-03-21 at 14 09 39@2x](https://github.com/laravel/nova-issues/assets/172966/2e4b5cfe-5095-457c-b17e-0de8369ca1cb)

Already exists for awhile.

**Comment by miagg** [comment]
> This functionality still depends on browser's default behavior.

This is not true. When `download` tag exists browser will always download the file instead of displaying it in browser. I've tried on numerous projects. It seems that there is something wrong with the js handler or with the inertia response.

Also I have no idea why this is converted to a discussion instead of an issue. Isn't the purpose of "downloadURL" to download? Or am I missing something?

**Reply by crynobone** [reply]
![CleanShot 2024-03-21 at 14 09 39@2x](https://github.com/laravel/nova-issues/assets/172966/2e4b5cfe-5095-457c-b17e-0de8369ca1cb)

Already exists for awhile.

**Reply by miagg** [reply]
@crynobone first of all this is not even the correct handler. This is the correct one:
```js
    if (data.download) {
      return emitResponseCallback(async () => {
        showActionResponseMessage(data)

        await nextTick(() => {
          let link = document.createElement('a')
          link.href = data.download
          link.download = data.name
          document.body.appendChild(link)
          link.click()
          document.body.removeChild(link)
        })
      })
    }
```
Note that the `data.name` here is the action name and NOT the file name which is totally wrong. However this is not the cause of the problem. Since you are depending on a "trick" to fake the download that is not working, this function is useless since `visit` already exists with the same functionality.

**Reply by crynobone** [reply, marked as answer]
![CleanShot 2024-03-21 at 17 02 07](https://github.com/laravel/nova-issues/assets/172966/ab33b967-33fd-4be9-93ae-3ff21f6b47d1)

There's another behavior that I had forgotten earlier. From: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a

**Reply by miagg** [reply]
Yes, this is exactly what I read and got me disappointed! In my case is cross-origin download. Serving qr-code images from subdomain (qr.example.net) instead of example.net. Once I use the original domain it downloads.

**Reply by miagg** [reply]
@crynobone should I open a new issue for the fact that the download filename is the name of the action?

I can't seem to work around it. I tried the following workaround but I cannot set a dynamic name based on the model.

**Originally (Problem)**
```php
Action::downloadURL('Download QR Code (SVG)', fn(Elabel $elabel) => $elabel->qr_svg_url)
    ->sole()->withoutConfirmation();
```
Downloads file as `Download QR Code (SVG).svg`

**Workaround (No dynamic name though)**
```php
Action::downloadURL("qrcode.svg", fn(Elabel $elabel) => $elabel->qr_svg_url)
    ->withName('Download QR Code (SVG)')
    ->sole()->withoutConfirmation(),
```
Downloads file as `qrcode.svg`

**Reply by miagg** [reply]
Nevermind, I submitted an issue #6318
